src/scaler/mod.rs

```diff
@@ -39,18 +60,363 @@ pub trait Scaler {
+            // will then ignore incoming events until all of the expected events have been received.
+            let expected_events = commands
+                .iter()
+                .filter_map(|cmd| cmd.corresponding_event(model_name));
+
+            // Only let other scalers know if we generated commands to take
+            // NOTE(brooksmtownsend): I tried to make the events iterator peekable and look at that instead
+            // but that resulted in a "higher ranked lifetime error". Weird.
+            if !commands.is_empty() {
+                trace!("Scaler generated commands, notifying other scalers to register expected events");
+                let data = serde_json::to_vec(&Notifications::RegisterExpectedEvents {
+                    name: model_name.to_owned(),
+                    scaler_id: self.scaler.id().to_owned(),
+                    triggering_event: Some(event.to_owned().try_into()?),
+                })?;
+
+                self.notifier
+                    .publish(data, Some(&self.notify_subject))
+                    .await?;
+            }
+            self.add_events(expected_events, false).await;
+            commands
+        };
+
+        Ok(commands)
+    }
+
+    #[instrument(level = "trace", skip_all, fields(scaler_id = %self.id()))]
+    async fn reconcile_internal(&self) -> Result<Vec<Command>> {
+        // If we're already in backoff, return an empty list
+        let current_event_count = self.event_count().await;
+        if current_event_count > 0 {
+            trace!(%current_event_count, "Scaler is backing off, not reconciling");
+            return Ok(Vec::with_capacity(0));
+        }
+        match self.scaler.reconcile().await {
+            // "Back off" scaler with expected corresponding events if the scaler generated commands
+            Ok(commands) if !commands.is_empty() => {
+                trace!("Reconcile generated commands, notifying other scalers to register expected events");
+                let data = serde_json::to_vec(&Notifications::RegisterExpectedEvents {
+                    name: self.model_name.to_owned(),
+                    scaler_id: self.scaler.id().to_owned(),
+                    triggering_event: None,
+                })?;
+                self.notifier
+                    .publish(data, Some(&self.notify_subject))
+                    .await?;
+                self.add_events(
+                    commands
+                        .iter()
+                        .filter_map(|command| command.corresponding_event(&self.model_name)),
+                    true,
+                )
+                .await;
+                Ok(commands)
+            }
+            Ok(commands) => {
+                trace!("Reconcile generated no commands, no need to register expected events");
+                Ok(commands)
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Sets a timed cleanup task to clear the expected events list after a timeout
```

Comment:
Why is this needed?

Reply:
We don't want to stay in a backoff state forever, this handles the case of a network interruption or missed events by eventually expiring them